cpu/rpx0xx/periph/uart.c

```diff
@@ -0,0 +1,201 @@
+ * @author      Justus Krebs <justus.krebs@st.ovgu.de>
+ * @author      Nick Weiler <nick.weiler@st.ovgu.de>
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ *
+ * @}
+ */
+
+#include "board.h"
+#include "bitarithm.h"
+#include "irq.h"
+#include "mutex.h"
+#include "periph/gpio.h"
+#include "periph/uart.h"
+#include "periph_cpu.h"
+#include "periph_conf.h"
+#include "io_reg.h"
+
+#include "assert.h"
+
+static uart_isr_ctx_t ctx[UART_NUMOF];
+
+void _irq_enable(uart_t uart)
+{
+    UART0_Type *dev = uart_config[uart].dev;
+    dev->UARTIMSC.reg = UART0_UARTIMSC_RXIM_Msk;
+    NVIC_EnableIRQ(uart_config[uart].irqn);
+}
+
+void _set_symbolrate(uart_t uart, uint32_t baud)
+{
+    assert(baud != 0);
+    UART0_Type *dev = uart_config[uart].dev;
+    uint32_t baud_rate_div = (8 * CLOCK_PERIPH / baud);
+    uint32_t baud_ibrd = baud_rate_div >> 7;
+    uint32_t baud_fbrd;
+
+    if (baud_ibrd == 0) {
+        baud_ibrd = 1;
+        baud_fbrd = 0;
+    }
+    else if (baud_ibrd >= 65535) {
+        baud_ibrd = 65535;
+        baud_fbrd = 0;
+    }
+    else {
+        baud_fbrd = ((baud_rate_div & 0x7f) + 1) / 2;
+    }
+
+    dev->UARTIBRD.reg = baud_ibrd;
+    dev->UARTFBRD.reg = baud_fbrd;
+
+    io_reg_atomic_set(&(dev->UARTLCR_H.reg), 0);
+}
+
+int uart_mode(uart_t uart, uart_data_bits_t data_bits, uart_parity_t uart_parity,
+              uart_stop_bits_t stop_bits)
+{
+    UART0_Type *dev = uart_config[uart].dev;
+
+    uint32_t config = (data_bits << UART0_UARTLCR_H_WLEN_Pos);
+
+    switch (uart_parity) {
+    case UART_PARITY_NONE:
+        break;
+    case UART_PARITY_EVEN:
+        config |= UART0_UARTLCR_H_EPS_Msk | UART0_UARTLCR_H_PEN_Msk;
+        break;
+    case UART_PARITY_ODD:
+        config |= UART0_UARTLCR_H_PEN_Msk;
+        break;
+    default:
+        return UART_NOMODE;
+    }
+
+    if (stop_bits == UART_STOP_BITS_2) {
+        config |= UART0_UARTLCR_H_STP2_Msk;
+    }
+
+    dev->UARTLCR_H.reg = config;
+
+    return UART_OK;
+}
+
+void uart_init_pins(uart_t uart)
+{
+    gpio_init(uart_config[uart].tx_pin, GPIO_OUT);
+    gpio_set_function_select(uart_config[uart].tx_pin, FUNCTION_SELECT_UART);
+    if (ctx[uart].rx_cb) {
+        gpio_init(uart_config[uart].rx_pin, GPIO_IN_PU);
+        gpio_set_function_select(uart_config[uart].rx_pin, FUNCTION_SELECT_UART);
+    }
+}
+
+void uart_deinit_pins(uart_t uart)
+{
+    gpio_reset_all_config(uart_config[uart].tx_pin);
+    SIO->GPIO_OE_CLR.reg = 1LU << uart_config[uart].tx_pin;
+    if (ctx[uart].rx_cb) {
+        gpio_reset_all_config(uart_config[uart].rx_pin);
+    }
+}
+
+void uart_poweron(uart_t uart)
+{
+    uint32_t reset_bit_mask = (uart) ? RESETS_RESET_uart1_Msk : RESETS_RESET_uart0_Msk;
+    periph_reset(reset_bit_mask);
+    periph_reset_done(reset_bit_mask);
+}
+
+void uart_poweroff(uart_t uart)
+{
+    uart_deinit_pins(uart);
+    periph_reset((uart) ? RESETS_RESET_uart1_Msk : RESETS_RESET_uart0_Msk);
+}
```

Comment:
Those are again wonderfully underspecified APIs. IMO, we should just drop `uart_poweron()`. Assume a UART is first initialized, then powered off, then power on again: Should it resume operation in the state before power off? Why should the driver need to sacrifice RAM to back up the configuration, when the user of the UART exactly knows all configuration details?

IMO, we should drop `uart_poweron()`. `uart_init()` powers on and initialized the UART, `uart_poweroff()` powers it off again. There is no use case for `uart_poweron()`.

Reply:
Oh, [it is actually documented](https://api.riot-os.org/group__drivers__periph__uart.html ):

> After initialization, the UART peripheral should be powered on and active. The UART can later be explicitly put to sleep and woken up by calling the uart_poweron() and uart_poweroff() functions. Once woken up using uart_poweron(), the UART should transparently continue it's previously configured operation.

Just not at the API of the functions. I still wonder why a low level driver is burdened with backing up the device state. Especially since in most case the compiler will not be able to optimize the RAM used for this out, even if it is never used.